File: src/main/java/org/samoxive/safetyjim/discord/commands/RoleCommand.java
package org.samoxive.safetyjim.discord.commands;

import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.*;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;
import org.jooq.DSLContext;
import org.samoxive.jooq.generated.Tables;
import org.samoxive.jooq.generated.tables.records.RolelistRecord;
import org.samoxive.safetyjim.discord.Command;
import org.samoxive.safetyjim.discord.DiscordBot;
import org.samoxive.safetyjim.discord.DiscordUtils;
import org.samoxive.safetyjim.discord.TextUtils;

import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public class RoleCommand extends Command {
    private String[] usages = {
            "role add <roleName> - adds a new self-assignable role",
            "role remove <roleName> - removes a self-assignable role",
    };

    @Override
    public String[] getUsages() {
        return usages;
    }

    @Override
    public boolean run(DiscordBot bot, GuildMessageReceivedEvent event, String args) {
        Scanner messageIterator = new Scanner(args);
        DSLContext database = bot.getDatabase();

        Member member = event.getMember();
        Message message = event.getMessage();
        Guild guild = event.getGuild();

        if (!messageIterator.hasNext()) {
            return true;
        }

        String subcommand = messageIterator.next();
        switch (subcommand) {
            case "add":
            case "remove":
                break;
            default:
                return true;
        }

        if (!member.hasPermission(Permission.ADMINISTRATOR)) {
            DiscordUtils.failMessage(bot, message, "You don't have enough permissions to execute this command! Required permission: Administrator");
            return false;
        }

        String roleName = TextUtils.seekScannerToEnd(messageIterator).toLowerCase();

        if (roleName.equals("")) {
            return true;
        }

        List<Role> matchingRoles = guild.getRoles()
                                       .stream()
                                       .filter((role) -> role.getName().toLowerCase().equals(roleName))
                                       .collect(Collectors.toList());

        if (matchingRoles.size() == 0) {
            DiscordUtils.failMessage(bot, message, "Could not find a role with specified name!");
            return false;
        }

        Role matchedRole = matchingRoles.get(0);

        if (subcommand.equals("add")) {
            RolelistRecord record = database.selectFrom(Tables.ROLELIST)
                    .where(Tables.ROLELIST.GUILDID.eq(guild.getId()))
                    .and(Tables.ROLELIST.ROLEID.eq(matchedRole.getId()))
                    .fetchAny();

            if (record == null) {
                record = database.newRecord(Tables.ROLELIST);
                record.setGuildid(guild.getId());
                record.setRoleid(matchedRole.getId());
                record.store();
                DiscordUtils.successReact(bot, message);
            } else {
                DiscordUtils.failMessage(bot, message, "Specified role is already in self-assignable roles list!");
                return false;
            }
        } else {
            RolelistRecord record = database.selectFrom(Tables.ROLELIST)
                    .where(Tables.ROLELIST.GUILDID.eq(guild.getId()))
                    .and(Tables.ROLELIST.ROLEID.eq(matchedRole.getId()))
                    .fetchAny();

            if (record == null) {
                DiscordUtils.failMessage(bot, message, "Specified role is not in self-assignable roles list!");
                return false;
            } else {
                record.delete();
                DiscordUtils.successReact(bot, message);
            }
        }

        return false;
    }
}

File: src/main/java/org/samoxive/safetyjim/discord/commands/RoleCommand.kt
package org.samoxive.safetyjim.discord.commands

import java.util.*
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent
import org.samoxive.safetyjim.database.RoleEntity
import org.samoxive.safetyjim.database.RolesTable
import org.samoxive.safetyjim.database.SettingsEntity
import org.samoxive.safetyjim.discord.*

class RoleCommand : Command() {
    override val usages = arrayOf("role add <roleName> - adds a new self-assignable role", "role remove <roleName> - removes a self-assignable role")

    override suspend fun run(bot: DiscordBot, event: GuildMessageReceivedEvent, settings: SettingsEntity, args: String): Boolean {
        val messageIterator = Scanner(args)

        val member = event.member!!
        val message = event.message
        val guild = event.guild

        if (!messageIterator.hasNext()) {
            return true
        }

        val subcommand = messageIterator.next()
        when (subcommand) {
            "add", "remove" -> {
            }
            else -> return true
        }

        if (!member.hasPermission(Permission.ADMINISTRATOR)) {
            message.failMessage("You don't have enough permissions to execute this command! Required permission: Administrator")
            return false
        }

        val roleName = messageIterator.seekToEnd().toLowerCase()

        if (roleName == "") {
            return true
        }

        val matchingRoles = guild.roles
                .filter { role -> role.name.toLowerCase() == roleName }

        if (matchingRoles.isEmpty()) {
            message.failMessage("Could not find a role with specified name!")
            return false
        }

        val matchedRole = matchingRoles[0]

        val record = RolesTable.fetchRole(guild, matchedRole)
        if (subcommand == "add") {
            if (record == null) {
                RolesTable.insertRole(
                        RoleEntity(
                                guildId = guild.idLong,
                                roleId = matchedRole.idLong
                        )
                )
                message.successReact()
            } else {
                message.failMessage("Specified role is already in self-assignable roles list!")
                return false
            }
        } else {
            if (record == null) {
                message.failMessage("Specified role is not in self-assignable roles list!")
                return false
            } else {
                RolesTable.deleteRole(record)
                message.successReact()
            }
        }

        return false
    }
}
